File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue <int> max_heap;
    priority_queue <int, vector <int>, greater <int>> min_heap;
    MedianFinder() {
        // if element bigger then push into max heap
        // check later maybe wrong
    }
    
    void addNum(int num) {
        if (max_heap.empty() || num <= max_heap.top()) max_heap.push(num); // check if < or >
        else min_heap.push(num);
        if (max_heap.size() > min_heap.size() + 1) { // have to check this 
            min_heap.push(max_heap.top());
            max_heap.pop();
        } else if (min_heap.size() > max_heap.size()) {
            max_heap.push(min_heap.top());
            min_heap.pop();
        }
    }
    // 2,3,4,5 --> 2, 3 4 5 -> 2, 3 and 4, 5 -> 2 and 3,4,5 even len
    // odd len 1,2,3 1 and 2,3 -> ok got it
    double findMedian() {
        if (max_heap.size() == min_heap.size()) {
            return (max_heap.top() + min_heap.top()) / 2.0;
        }
        return max_heap.top();
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.top = []
        self.bottom = []

    def addNum(self, num: int) -> None:
        heapq.heappush(self.bottom, -1 * num)
        heapq.heappush(self.top, -1 * heapq.heappop(self.bottom))

        if len(self.top) > len(self.bottom):
            heapq.heappush(self.bottom, -1 * heapq.heappop(self.top))

    def findMedian(self) -> float:
        if len(self.bottom) == len(self.top):
            return (self.top[0] - self.bottom[0]) / 2

        return -1 * self.bottom[0]
        
# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()